eva/executor/create_index_executor.py

```diff
@@ -0,0 +1,220 @@
+                logical_to_row_id[num_rows] = row_id
+
+                num_rows += 1
+
+            # Get index IO list.
+            io_list = self._get_index_io_list(input_dim)
+
+            # Build secondary index maaping.
+            secondary_index_df_metadata = self._create_secondary_index_table()
+            storage_engine = StorageEngine.factory(secondary_index_df_metadata)
+            storage_engine.create(secondary_index_df_metadata)
+
+            # Write mapping.
+            logical_id, row_id = list(logical_to_row_id.keys()), list(
+                logical_to_row_id.values()
+            )
+            secondary_index = Batch(
+                pd.DataFrame(data={"logical_id": logical_id, "row_id": row_id})
+            )
+            storage_engine.write(secondary_index_df_metadata, secondary_index)
+
+            # Persist index.
+            faiss.write_index(index, self._get_index_save_path())
+
+            # Save to catalog.
+            catalog_manager.create_index(
+                self.node.name,
+                self._get_index_save_path(),
+                self.node.index_type,
+                io_list,
+            )
+
+            yield Batch(
+                pd.DataFrame(
+                    [f"Index {self.node.name} successfully added to the database."]
+                )
+            )
+        except Exception as e:
+            # Roll back in reverse order.
+            # Delete on-disk index.
+            if os.path.exists(self._get_index_save_path()):
+                os.remove(self._get_index_save_path())
+
+            # Drop secondary index table.
+            secondary_index_tb_name = "secondary_index_{}_{}".format(
+                self.node.index_type, self.node.name
+            )
+            if catalog_manager.check_table_exists(
+                None,
+                secondary_index_tb_name,
+            ):
+                secondary_index_metadata = catalog_manager.get_dataset_metadata(
+                    None, secondary_index_tb_name
+                )
+                storage_engine = StorageEngine.factory(secondary_index_metadata)
+                storage_engine.drop(secondary_index_metadata)
+                catalog_manager.drop_dataset_metadata(
+                    None,
+                    secondary_index_tb_name,
+                )
+
+            # Throw exception back to user.
+            raise ExecutorError(str(e))
+
+    def _get_index_save_path(self):
+        return str(
+            EVA_DEFAULT_DIR
+            / INDEX_DIR
+            / Path("{}_{}.index".format(self.node.index_type, self.node.name))
+        )
+
+    def _get_index_io_list(self, input_dim):
+        # Input dimension is inferred from the actual feature.
+        catalog_manager = CatalogManager()
+        input_index_io = catalog_manager.index_io(
+            "input_feature",
+            ColumnType.NDARRAY,
+            NdArrayType.FLOAT32,
+            [Dimension.ANYDIM, input_dim],
+            True,
```

Comment:
After some brainstorming, I think it is better to just store those information in `IndexMetadata` as opposed to `IndexIO`. `IndexIO` can be both input and output, so it does not make sense to store those info for output as well. What do you guys think?

Reply:
Yes, makes sense. Quick question, how do you plan to handle multiple-column cases?

Reply:
Fix this. Ready for another review. Only thing I left out is udf reference. Since this PR assumes the feature vector has already been materialized in a table, we can add that reference in future PRs.